### ISSUE TITLE
Bot websocket closes with code 1000 when under http load

### ISSUE THREAD

**Issue by yesimon** [Rapptz/discord.py#229]
My bot has been dying with websocket code 1000 (normal close). Previously the bot would run forever, but after interfacing with another API and doing aiohttp get requests on the same event loop (10 req/10 s), I would get a close after about 15 mins. Without load it runs forever. Is there some kind of interference between aiohttp requests and websockets on the same event loop? Secondly, shouldn't the websocket never close with code 1000?

I the below I added a bit of logging to `discord.py` to validate that it is indeed exiting with code 1000.

```
2016/05/28 12:05:51 AM:DEBUG:websockets.protocol: client >> Frame(fin=True, opcode=1, data=b'{"op":1,"d":2}')
2016/05/28 12:05:51 AM:DEBUG:websockets.protocol: client << Frame(fin=True, opcode=8, data=b'\x03\xe8')
2016/05/28 12:05:51 AM:DEBUG:websockets.protocol: client >> Frame(fin=True, opcode=8, data=b'\x03\xe8')
2016/05/28 12:05:51 AM:DEBUG:asyncio: <_SelectorSocketTransport fd=13 read=polling write=<idle, bufsize=0>> received EOF
2016/05/28 12:05:51 AM:DEBUG:asyncio: <asyncio.sslproto.SSLProtocol object at 0x7fa0a65a76a0> received EOF
2016/05/28 12:05:51 AM:ERROR:discord.client: WS yielding failed: WebSocket connection is closed: code = 1000, no reason.
Traceback (most recent call last):
  File "/home/yesimon/src/discord.py/discord/gateway.py", line 346, in poll_event
    msg = yield from self.recv()
  File "/home/yesimon/.pyenv/versions/3.5.1/lib/python3.5/asyncio/coroutines.py", line 105, in __next__
    return self.gen.send(None)
  File "/home/yesimon/.pyenv/versions/3.5.1/lib/python3.5/site-packages/websockets/protocol.py", line 296, in recv
    raise ConnectionClosed(self.close_code, self.close_reason)
websockets.exceptions.ConnectionClosed: WebSocket connection is closed: code = 1000, no reason.

The above exception was the direct cause of the following exception:

Traceback (most recent call last):
  File "/home/yesimon/src/discord.py/discord/client.py", line 403, in connect
    yield from self.ws.poll_event()
  File "/home/yesimon/.pyenv/versions/3.5.1/lib/python3.5/asyncio/coroutines.py", line 105, in __next__
    return self.gen.send(None)
  File "/home/yesimon/src/discord.py/discord/gateway.py", line 353, in poll_event
    raise ConnectionClosed(e) from e
discord.errors.ConnectionClosed: WebSocket connection is closed: code = 1000, no reason.
2016/05/28 12:05:51 AM:ERROR:discord.client: ws is closed
2016/05/28 12:05:51 AM:DEBUG:asyncio: Close <_UnixSelectorEventLoop running=False closed=False debug=True>
tail: discord.log: file truncated
2016/05/28 12:05:51 AM:ERROR:asyncio: Task was destroyed but it is pending!
```

**Comment by Taiiwo** [comment]
Getting this error after about 2-3 minutes of running my code. Happens consistently without me making any of my own HTTP requests.

My bot runs a `while True` in a message event handler, could that be why it's timing out? Because the event handlers stop the bot from pinging if they block execution?

**Comment by Titouan-cpu** [comment]
I still have this problem. Is there any solution yet ?

**Comment by Taiiwo** [comment]
Looking back on this knowing a bit more about asyncio, it seems the solution is quite obvious. I assume that was why it was closed without explanation, but I'll clear it up in case any other newcomers have the same issue. Basically, if you're blocking the execution for more than 240 seconds, discord.py will break. In OPs case, his web requests were taking so long that it took over 240 seconds for the event loop to get back around to discord requests, thus closing the connection. In my case more simply, I was unaware you weren't supposed to block execution flow inside of a coroutine, and that it would suffocate the discord connection.

The solution is to make sure you're not blocking the execution flow, or adding _too many_ things to the event loop. This will be specific to your application, but in OPs case, he could probably put an `asyncio.sleep(0)` between every few tasks they add to the loop, so as to give discord.py a chance to put some requests in. Or run the web requests in a thread..

**Comment by yesimon** [comment]
@Taiiwo Where do you get the magic number 240s? For the current code I only see some references to timeout=180s in the shard.py https://github.com/Rapptz/discord.py/blob/13a3f760e6de8aa251f06bbe2a746e5f92deafd2/discord/shard.py#L84.

It seems like the current version of the code is a bit different like accepting code 1000s sent from discord https://github.com/Rapptz/discord.py/commit/5ce88c8a3f7457384e064c2f3ce030dbc3bae9e0 changing default WS close to 4000 instead of 1000 https://github.com/Rapptz/discord.py/blob/baffe564e52c409b9eb85f5d52e309948f0611d9/docs/whats_new.rst#L22. I think the original bug report is obsolete.

**Comment by gogurtenjoyer** [comment]
Just a quick reminder: this issue was closed 3.5 years ago - there's no need to add comments to such old issues.

Also, to correct some misinformation: while blocking is bad, I'm not sure it had anything to do with the original issue and the solutions provided above are in fact not solutions. I recommend [reading up about asyncio](https://docs.python.org/3/library/asyncio.html) and perhaps [aiohttp ](https://docs.aiohttp.org/en/stable/client.html)as a solution for making asynchronous web requests that do not block the bot's event loop. Threading and 0 length sleeps are not what you want here.

**Comment by caio-vinicius** [comment]
Thank you guys.
